// bn-schema-isbd/index.js
module.exports = {
    directory: __dirname,
    prefix: 'isbd',
    fields: {
        titleproper: {
            type: 'text',
            label: 'Title proper',
            css: 'font-weight: bold; font-size: 120%;'
        },
        titleremainder: {
            type: 'text',
            label: 'Remainder of title',
            css: 'font-weight: bold;'
        },
        edition: {
            type: 'text',
            label: 'Edition',
            css: ''
        },
        source: {
            type: 'text',
            label: 'Source'
        },
        publisher: {
            type: 'text',
            label: 'Publisher'
        },
        titlevariant: {
            type: 'text',
            label: 'Other title'
        },
        titletranslated: {
            type: 'text',
            label: 'Translated title'
        },
        uri: {
            type: 'property',
            unique: false,
            multivalue: true,
            datatype: 'String',
            label: 'URI'
        },
        note: {
            type: 'text',
            label: 'Note'
        },
        author: {
            type: 'text',
            label: 'Author'
        },
        host: {
            type: 'text',
            label: 'Host item'
        },
        subject: {
            type: 'text',
            label: 'Subject'
        },
        place: {
            type: 'text',
            label: 'Place of publication'
        },
        callnum: {
            type: 'text',
            label: 'Call number'
        },
        year: {
            type: 'property',
            datatype: 'Integer',
            label: 'Publication date'
        },
        barcode: {
            type: 'property',
            datatype: 'String',
            label: 'Barcode'
        },
        series: {
            type: 'text',
            label: 'Series'
        },
        adate: {
            type: 'property',
            datatype: 'String',
            label: 'Auction date'
        },
        author_e: {
            type: 'edge',
            unidirected: false,
            label: 'Author (edge)'
        },
        subject_e: {
            type: 'edge',
            unidirected: false,
            label: 'Subject (edge)'
        },
        place_e: {
            type: 'edge',
            unidirected: false,
            label: 'Place (edge)'
        },
        year_e: {
            type: 'edge',
            unidirected: false,
            label: 'Year (edge)'
        },
        series_e: {
            type: 'edge',
            unidirected: false,
            label: 'Series (edge)'
        },
    },
    facets: {
        author_e: {
            outlabel: 'By',
            inlabel: 'Wrote',
            facetlabel: 'Author',
            field: 'author_e'
        },
        subject_e: {
            outlabel: 'About',
            inlabel: 'Topic of',
            facetlabel: 'Subject',
            field: 'subject_e'
        },
        place_e: {
            outlabel: 'Published in',
            inlabel: 'Place of',
            facetlabel: 'Place of publication',
            field: 'place_e'
        },
        year_e: {
            outlabel: 'Published in',
            inlabel: 'Year of',
            facetlabel: 'Date of publication',
            field: 'year_e'
        },
        series_e: {
            outlabel: 'Part of series',
            inlabel: 'Series of',
            facetlabel: 'Series',
            field: 'series_e'
        }
    },
    static_relevance_bumps: {
        recordclass: {
            biblio: 1,
            auth: 0.5
        }
    },
    templates: [
        'marc21'
    ]
};
