// bn-schema-isbd/index.js
module.exports = {
    prefix: 'isbd',
    fields: {
        titleproper: {
            type: 'text',
            label: 'Title proper',
            css: 'font-weight: bold; font-size: 120%;'
        },
        titleremainder: {
            type: 'text',
            label: 'Remainder of title',
            css: 'font-weight: bold;'
        },
        edition: {
            type: 'text',
            label: 'Edition',
            css: ''
        },
        source: {
            type: 'text',
            label: 'Source'
        },
        publisher: {
            type: 'text',
            label: 'Publisher'
        },
        titlevariant: {
            type: 'text',
            label: 'Other title'
        },
        titletranslated: {
            type: 'text',
            label: 'Translated title'
        },
        uri: {
            type: 'property',
            unique: false,
            multivalue: true,
            datatype: 'String',
            label: 'URI'
        },
        note: {
            type: 'text',
            label: 'Note'
        },
        author: {
            type: 'text',
            label: 'Author'
        },
        subject: {
            type: 'text',
            label: 'Subject'
        },
        author_e: {
            type: 'edge',
            unidirected: false,
            label: 'Author (edge)'
        },
        subject_e: {
            type: 'edge',
            unidirected: false,
            label: 'Subject (edge)'
        },
    },
    facets: {
        author_e: {
            outlabel: 'By',
            inlabel: 'Wrote',
            facetlabel: 'Author',
            field: 'author_e'
        },
        subject_e: {
            outlabel: 'About',
            inlabel: 'Topic of',
            facetlabel: 'Subject',
            field: 'subject_e'
        }
    }
};
