// bn-schema-ericthesaurus/index.js
module.exports = {
    prefix: 'eric',
    fields: {
        scopenote: {
            type: 'text',
        },
        preferred: {
            type: 'edge',
            unidirected: false
        },
        synonym: {
            type: 'inverseedge',
            edge: 'preferred'
        },
        broader: {
            type: 'edge',
            unidirected: false
        },
        narrower: {
            type: 'inverseedge',
            edge: 'broader'
        },
        related: {
            type: 'biedge',
            unidirected: false
        }
    },
    facets: {
        broader: {
            outlabel: 'Broader term',
            inlabel: 'Narrower term',
            facetlabel: 'Broader term',
            field: 'broader'
        },
        preferred: {
            outlabel: 'Preferred term',
            inlabel: 'Used for',
            facetlabel: 'Preferred term',
            field: 'preferred'
        },
        related: {
            outlabel: 'Related term',
            inlabel: 'Related term',
            facetlabel: 'Related term',
            field: 'related'
        }
    }
};
