// bn-schema-common/index.js
module.exports = {
    prefix: 'bn',
    fields: {
        keyword: {
            type: 'text',
            unique: false,
            multivalue: false,
            system: true
        },
        key: {
            type: 'property',
            datatype: 'String',
            unique: true,
            multivalue: false,
            system: true
        },
        data: {
            type: 'text',
            system: true
        },
        format: {
            type: 'property',
            datatype: 'String',
            unique: false,
            multivalue: false,
            system: true
        },
        recordtype: {
            type: 'edge',
            unidirected: true,
            system: true
        },
        recordclass: {
            type: 'property',
            datatype: 'String',
            unique: false,
            multivalue: false,
            system: true
        },
        model: {
            type: 'property',
            datatype: 'String',
            unique: false,
            multivalue: false,
            system: true
        },
        linkbrowse: {
            type: 'dbcallback',
            system: true
        },
        vorder: {
            type: 'property',
            datatype: 'Integer',
            unique: false,
            multivalue: false,
            system: true
        }
    },
    facets: {
        recordtype: {
            outlabel: 'Is a',
            inlabel: 'Example of',
            facetlabel: 'Record type',
            field: 'recordtype'
        }
    }
};
