// bn-schema-eric/index.js
module.exports = {
    directory: __dirname,
    prefix: 'eric',
    fields: {
        creator: {
            type: 'edge',
            unidirected: false,
            label: 'Creator'
        },
        wrote: {
            type: 'inverseedge',
            edge: 'creator'
        },
        title: {
            type: 'text',
            label: 'Title'
        },
        source: {
            type: 'text',
            label: 'Source'
        },
        citation: {
            type: 'text',
            label: 'Citation'
        },
        description: {
            type: 'text',
            label: 'Description'
        },
        publisher: {
            type: 'text',
            label: 'Publisher'
        },
        uri: {
            type: 'property',
            unique: false,
            multivalue: true,
            label: 'URI'
        },
        subject: {
            type: 'edge',
            label: 'Subject'
        }
    },
    facets: {
        author: {
            outlabel: 'By',
            inlabel: 'Wrote',
            facetlabel: 'Author',
            field: 'creator'
        },
        subject: {
            outlabel: 'About',
            inlabel: 'Topic of',
            facetlabel: 'Subject'
        }
    },
    static_relevance_bumps: {
        recordclass: {
            biblio: 2,
            term: 0.5,
            synonym: -0.5
        }
    },
    templates: [
        'eric'
    ]
};
