// bngraphstore/index.js
"use strict";
var environment = require('../../lib/environment'),
    g = require('gremlin');

var db;
var engine;

module.exports = function graph(opts) {
    engine = environment.graphstore.engine;
    if (typeof opts !== 'undefined') {
        engine = opts.engine || engine;
        if (typeof db !== 'undefined') {
            db.shutdownSync();
            db = undefined;
        }
    }
    if (typeof db === 'undefined') {
        connect();
    }
    return g;
};

module.exports.getDB = function () {
    return db;
};

module.exports.getEngine = function () {
    return engine;
};

module.exports.textSearch = function (query) {
    if (engine === 'titan') {
        var Text = g.java.import('com.thinkaurelius.titan.core.attribute.Text');
        var temp =  g.start(db.querySync().hasSync('data', Text.CONTAINS, query).verticesSync());
        return temp;
    } else {
        return g.V().filter("{it.data?.count('" + query + "') >= 1}");
    }
};

module.exports.autocommit = true;

function connect() {
    engine = engine || environment.graphstore.engine;

    if (engine === 'titan') {
        //Get a reference to Titan specific Enum
        var BaseConfiguration = g.java.import('org.apache.commons.configuration.BaseConfiguration');
        var TitanFactory = g.java.import('com.thinkaurelius.titan.core.TitanFactory');
        var Direction = g.Direction,
            Type = g.ClassTypes;
        var UniqCon = g.java.import("com.thinkaurelius.titan.core.TypeMaker$UniquenessConsistency");

        var gconf = new BaseConfiguration();
        for (var property in environment.graphstore[engine]) {
            gconf.setPropertySync(property, environment.graphstore[engine][property]);
        }
        db = TitanFactory.openSync(gconf);

        for (var name in environment.indexes) {
            var index = environment.indexes[name];
            try {
                switch (index.type) {
                    case 'edge':
                        if (index.unidirected) {
                            db.makeTypeSync().nameSync(name).unidirectedSync().makeEdgeLabelSync();
                        } else {
                            db.makeTypeSync().nameSync(name).makeEdgeLabelSync();
                        }
                        break;
                    case 'property':
                        if (index.unique && !index.multivalue) {
                            db.makeTypeSync().nameSync(name).dataTypeSync(Type[index.datatype].class)
                                .indexedSync(Type.Vertex.class)
                                .uniqueSync(Direction.BOTH, UniqCon.LOCK).makePropertyKeySync();
                        } else if (index.unique) {
                            db.makeTypeSync().nameSync(name).dataTypeSync(Type[index.datatype].class)
                                .indexedSync(Type.Vertex.class).uniqueSync(Direction.IN).makePropertyKeySync();
                        } else if (!index.multivalue) {
                            db.makeTypeSync().nameSync(name).dataTypeSync(Type[index.datatype].class)
                                .indexedSync(Type.Vertex.class).uniqueSync(Direction.OUT).makePropertyKeySync();
                        } else {
                            db.makeTypeSync().nameSync(name).dataTypeSync(Type[index.datatype].class)
                                .indexedSync(Type.Vertex.class).makePropertyKeySync();
                        }
                        break;
                    case 'text':
                        db.makeTypeSync().nameSync(name).dataTypeSync(Type.String.class)
                            .indexedSync("search", Type.Vertex.class)
                            .uniqueSync(Direction.OUT).makePropertyKeySync();
                        break;
                }
            } catch (e) {
            }
        }
    } else if (engine === 'orient') {
        var OrientGraph = g.java.import('com.tinkerpop.blueprints.impls.orient.OrientGraph');
        db = new OrientGraph(environment.graphstore[engine].path, environment.graphstore[engine].username, environment.graphstore[engine].password);
    } else if (engine === 'tinker') {
        var TinkerGraph = g.java.import("com.tinkerpop.blueprints.impls.tg.TinkerGraph");
        if (environment.graphstore[engine].path === null) {
            db = new TinkerGraph();
        } else {
            db = new TinkerGraph(environment.graphstore[engine].path);
        }
        db.commitSync = function () {
        };
    } else if (engine === 'neo4j') {
        var Neo4jGraph = g.java.import("com.tinkerpop.blueprints.impls.neo4j.Neo4jGraph");
        db = new Neo4jGraph(environment.graphstore[engine].path);
    }
    process.on('exit', interruptHandler);
    process.on('SIGINT', function () {
        process.exit();
    });

    g.SetGraph(db);
}

function interruptHandler() {
    //db.commitSync();
    db.shutdownSync();
}
